#include "mainwindow.h"
#include "./ui_mainwindow.h"
#include <QPainter>
#include <QMouseEvent>
#include <QResizeEvent>
#include <QRandomGenerator>
#include <QMessageBox>
#include <QDateTime>
#include <QDebug>
#include <QApplication>
#include <QScreen>
#include <QLinearGradient>

/**
 * @brief æ„é€ å‡½æ•°
 * @param parent çˆ¶çª—å£
 */
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , m_hasCalculated(false)
    , m_gridSize(50)
    , m_showGridMode(false)  // é»˜è®¤æ˜¾ç¤ºXYåæ ‡å›¾
{
    ui->setupUi(this);
    initializeUI();
}

/**
 * @brief ææ„å‡½æ•°
 */
MainWindow::~MainWindow()
{
    delete ui;
}

/**
 * @brief åˆå§‹åŒ–ç•Œé¢
 */
void MainWindow::initializeUI()
{
    // è®¾ç½®çª—å£æ ‡é¢˜
    setWindowTitle("é‚®å±€é€‰å€é—®é¢˜");
    
    // è®¾ç½®çª—å£å¤§å°å’Œæœ€å°å°ºå¯¸
    resize(1400, 900);
    setMinimumSize(1200, 700);
    
    // è¿æ¥ä¿¡å·å’Œæ§½
    connect(ui->btnCalculate, &QPushButton::clicked, this, &MainWindow::calculateOptimalLocation);
    connect(ui->btnClear, &QPushButton::clicked, this, &MainWindow::clearAllAreas);
    connect(ui->btnRandom, &QPushButton::clicked, this, &MainWindow::generateRandomAreas);
    
    // åˆå§‹åŒ–ç»˜å›¾åŒºåŸŸ - frameæ¡†ç°åœ¨åœ¨æœ€å·¦ä¾§ï¼Œç»˜å›¾åŒºåŸŸåœ¨å³ä¾§
    int padding = 20;
    int frameWidth = ui->frame->width();
    int availableWidth = this->width() - frameWidth - padding * 2; // å‡å»æ§åˆ¶é¢æ¿å®½åº¦å’Œè¾¹è·
    int availableHeight = this->height() - 100; // å‡å»çŠ¶æ€æ å’Œè¾¹è·
    
    // è®¾ç½®ç»˜å›¾åŒºåŸŸå°ºå¯¸ï¼Œç¡®ä¿ä¸å°äºæœ€å°å°ºå¯¸
    int drawingWidth = qMax(800, availableWidth);
    int drawingHeight = qMax(600, availableHeight);
    
    // ç»˜å›¾åŒºåŸŸä½äºframeæ¡†å³ä¾§ï¼Œä¿æŒé€‚å½“é—´è·
    m_drawingArea = QRect(frameWidth + padding, padding, drawingWidth, drawingHeight);
    m_logicalBounds = QRect(-200, -200, 400, 400); // é€»è¾‘åæ ‡èŒƒå›´
    m_gridSize = 50; // ç½‘æ ¼å°ºå¯¸
    
    // è®¡ç®—ç¼©æ”¾æ¯”ä¾‹
    m_scaleX = static_cast<double>(m_drawingArea.width()) / m_logicalBounds.width();
    m_scaleY = static_cast<double>(m_drawingArea.height()) / m_logicalBounds.height();
    
    // æ›´æ–°çŠ¶æ€æ 
    updateStatusBar();
}

/**
 * @brief é‡ç»˜äº‹ä»¶å¤„ç†å‡½æ•°
 * @param event é‡ç»˜äº‹ä»¶
 */
void MainWindow::paintEvent(QPaintEvent *event)
{
    Q_UNUSED(event)
    
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    
    // ç»˜åˆ¶æ§åˆ¶é¢æ¿åŒºåŸŸèƒŒæ™¯
    painter.fillRect(ui->frame->geometry(), QColor(248, 249, 250));
    
    // ç»˜åˆ¶ç»˜å›¾åŒºåŸŸèƒŒæ™¯ - æ¸å˜èƒŒæ™¯
    QLinearGradient gradient(m_drawingArea.topLeft(), m_drawingArea.bottomRight());
    gradient.setColorAt(0, QColor(255, 255, 255));
    gradient.setColorAt(1, QColor(245, 245, 245));
    painter.fillRect(m_drawingArea, gradient);
    
    // ç»˜åˆ¶ç»˜å›¾åŒºåŸŸè¾¹æ¡†
    painter.setPen(QPen(QColor(222, 226, 230), 2));
    painter.drawRect(m_drawingArea.adjusted(0, 0, -1, -1));
    
    // ä¿å­˜ç»˜å›¾çŠ¶æ€
    painter.save();
    
    // è®¾ç½®ç»˜å›¾åŒºåŸŸä¸ºm_drawingArea
    painter.translate(m_drawingArea.topLeft());
    
    // æ ¹æ®æ˜¾ç¤ºæ¨¡å¼é€‰æ‹©ç»˜åˆ¶æ–¹å¼
    if (m_showGridMode) {
        // ç½‘æ ¼å›¾æ¨¡å¼ï¼šåªç»˜åˆ¶ç®€å•çš„ç½‘æ ¼ï¼Œä¸æ˜¾ç¤ºåæ ‡è½´å’Œæ ‡ç­¾
        drawSimpleGrid(painter, QRect(0, 0, m_drawingArea.width(), m_drawingArea.height()));
        
        // åœ¨ç½‘æ ¼å›¾æ¨¡å¼ä¸‹ä¹Ÿç»˜åˆ¶å±…æ°‘å°åŒºå’Œé‚®å±€
        drawResidentialAreas(painter);
        drawPostOffice(painter);
        
        // å¦‚æœå·²è®¡ç®—æœ€ä¼˜ä½ç½®ï¼Œç»˜åˆ¶è¿æ¥çº¿
        if (m_hasCalculated) {
            drawConnections(painter);
        }
    } else {
        // XYåæ ‡å›¾æ¨¡å¼ï¼šç»˜åˆ¶å®Œæ•´çš„åæ ‡ç³»ç»Ÿ
        drawGrid(painter, QRect(0, 0, m_drawingArea.width(), m_drawingArea.height()));
        
        // ç»˜åˆ¶å±…æ°‘å°åŒº
        drawResidentialAreas(painter);
        
        // ç»˜åˆ¶é‚®å±€ä½ç½®
        drawPostOffice(painter);
        
        // å¦‚æœå·²è®¡ç®—æœ€ä¼˜ä½ç½®ï¼Œç»˜åˆ¶è¿æ¥çº¿ï¼ˆåœ¨å°åŒºä¸Šæ–¹ï¼‰
        if (m_hasCalculated) {
            drawConnections(painter);
        }
    }
    
    // æ¢å¤ç»˜å›¾çŠ¶æ€
    painter.restore();
}

/**
 * @brief ç»˜åˆ¶åæ ‡ç½‘æ ¼
 * @param painter ç»˜å›¾å¯¹è±¡
 * @param contentRect å†…å®¹åŒºåŸŸ
 */
void MainWindow::drawGrid(QPainter &painter, const QRect &contentRect)
{
    painter.save();
    
    // è®¾ç½®ç»˜å›¾åŒºåŸŸä¸ºå†…å®¹åŒºåŸŸ
    painter.setClipRect(contentRect);
    
    // ç»˜åˆ¶ä¸»ç½‘æ ¼çº¿ - æ›´ç²—æ›´æ˜æ˜¾çš„ç½‘æ ¼ï¼Œå¢å¼ºå°çª—å£ä¸‹çš„å¯è§æ€§
    QPen mainGridPen(QColor(180, 180, 180, 180)); // æ›´æ·±çš„é¢œè‰²ï¼Œæ›´é«˜çš„ä¸é€æ˜åº¦
    mainGridPen.setWidth(2); // å¢åŠ çº¿å®½
    painter.setPen(mainGridPen);
    
    // ç»˜åˆ¶ä¸»ç½‘æ ¼çº¿ï¼ˆæ¯éš”ä¸€ä¸ªç½‘æ ¼ç»˜åˆ¶æ›´ç²—çš„çº¿ï¼‰
    for (int x = m_logicalBounds.left(); x <= m_logicalBounds.right(); x += m_gridSize * 2) {
        QPoint start = logicalToScreen(QPoint(x, m_logicalBounds.top()));
        QPoint end = logicalToScreen(QPoint(x, m_logicalBounds.bottom()));
        painter.drawLine(start, end);
    }
    
    for (int y = m_logicalBounds.top(); y <= m_logicalBounds.bottom(); y += m_gridSize * 2) {
        QPoint start = logicalToScreen(QPoint(m_logicalBounds.left(), y));
        QPoint end = logicalToScreen(QPoint(m_logicalBounds.right(), y));
        painter.drawLine(start, end);
    }
    
    // ç»˜åˆ¶æ¬¡ç½‘æ ¼çº¿ - è¾ƒç»†çš„ç½‘æ ¼çº¿
    QPen subGridPen(QColor(200, 200, 200, 150));
    subGridPen.setWidth(1);
    painter.setPen(subGridPen);
    
    for (int x = m_logicalBounds.left(); x <= m_logicalBounds.right(); x += m_gridSize) {
        if (x % (m_gridSize * 2) != 0) { // è·³è¿‡ä¸»ç½‘æ ¼çº¿
            QPoint start = logicalToScreen(QPoint(x, m_logicalBounds.top()));
            QPoint end = logicalToScreen(QPoint(x, m_logicalBounds.bottom()));
            painter.drawLine(start, end);
        }
    }
    
    for (int y = m_logicalBounds.top(); y <= m_logicalBounds.bottom(); y += m_gridSize) {
        if (y % (m_gridSize * 2) != 0) { // è·³è¿‡ä¸»ç½‘æ ¼çº¿
            QPoint start = logicalToScreen(QPoint(m_logicalBounds.left(), y));
            QPoint end = logicalToScreen(QPoint(m_logicalBounds.right(), y));
            painter.drawLine(start, end);
        }
    }
    
    // ç»˜åˆ¶åæ ‡è½´ - æ›´ç²—æ›´æ˜æ˜¾çš„åæ ‡è½´
    QPen axisPen(QColor(52, 73, 94));
    axisPen.setWidth(3);
    painter.setPen(axisPen);
    
    QPoint origin = logicalToScreen(QPoint(0, 0));
    QPoint xAxisEnd = logicalToScreen(QPoint(m_logicalBounds.right(), 0));
    QPoint yAxisEnd = logicalToScreen(QPoint(0, m_logicalBounds.bottom()));
    
    painter.drawLine(origin, xAxisEnd);
    painter.drawLine(origin, yAxisEnd);
    
    // ç»˜åˆ¶ç½‘æ ¼æ ‡ç­¾ - å¢å¼ºå°çª—å£ä¸‹çš„å¯è¯»æ€§
    painter.setPen(QColor(100, 100, 100, 200));
    QFont font = painter.font();
    font.setPointSize(8);
    font.setBold(true);
    painter.setFont(font);
    
    // ç»˜åˆ¶Xè½´ç½‘æ ¼æ ‡ç­¾
    for (int x = m_logicalBounds.left(); x <= m_logicalBounds.right(); x += m_gridSize * 2) {
        if (x != 0) { // è·³è¿‡åŸç‚¹
            QPoint labelPos = logicalToScreen(QPoint(x, 0));
            painter.drawText(labelPos + QPoint(-10, 15), QString::number(x));
        }
    }
    
    // ç»˜åˆ¶Yè½´ç½‘æ ¼æ ‡ç­¾
    for (int y = m_logicalBounds.top(); y <= m_logicalBounds.bottom(); y += m_gridSize * 2) {
        if (y != 0) { // è·³è¿‡åŸç‚¹
            QPoint labelPos = logicalToScreen(QPoint(0, y));
            painter.drawText(labelPos + QPoint(-25, 5), QString::number(y));
        }
    }
    
    // ç»˜åˆ¶åæ ‡æ ‡ç­¾
    painter.setPen(QColor(44, 62, 80));
    font.setPointSize(12);
    font.setBold(true);
    painter.setFont(font);
    painter.drawText(xAxisEnd + QPoint(10, 10), "Xè½´");
    painter.drawText(yAxisEnd + QPoint(10, 10), "Yè½´");
    
    // ç»˜åˆ¶åŸç‚¹æ ‡ç­¾
    font.setPointSize(10);
    painter.setFont(font);
    painter.drawText(origin + QPoint(-15, 15), "O(0,0)");
    
    painter.restore();
}

/**
 * @brief ç»˜åˆ¶ç®€å•ç½‘æ ¼ï¼ˆç½‘æ ¼å›¾æ¨¡å¼ï¼‰
 * @param painter ç»˜å›¾å¯¹è±¡
 * @param contentRect å†…å®¹åŒºåŸŸ
 */
void MainWindow::drawSimpleGrid(QPainter &painter, const QRect &contentRect)
{
    painter.save();
    
    // è®¾ç½®ç»˜å›¾åŒºåŸŸä¸ºå†…å®¹åŒºåŸŸ
    painter.setClipRect(contentRect);
    
    // ç»˜åˆ¶ç®€å•çš„ç½‘æ ¼çº¿ - æ›´ç»†æ›´ç®€æ´çš„ç½‘æ ¼
    QPen gridPen(QColor(220, 220, 220, 150)); // æµ…ç°è‰²ç½‘æ ¼
    gridPen.setWidth(1);
    painter.setPen(gridPen);
    
    // ç»˜åˆ¶æ°´å¹³ç½‘æ ¼çº¿
    for (int y = m_logicalBounds.top(); y <= m_logicalBounds.bottom(); y += m_gridSize) {
        QPoint start = logicalToScreen(QPoint(m_logicalBounds.left(), y));
        QPoint end = logicalToScreen(QPoint(m_logicalBounds.right(), y));
        painter.drawLine(start, end);
    }
    
    // ç»˜åˆ¶å‚ç›´ç½‘æ ¼çº¿
    for (int x = m_logicalBounds.left(); x <= m_logicalBounds.right(); x += m_gridSize) {
        QPoint start = logicalToScreen(QPoint(x, m_logicalBounds.top()));
        QPoint end = logicalToScreen(QPoint(x, m_logicalBounds.bottom()));
        painter.drawLine(start, end);
    }
    
    // ç»˜åˆ¶æ›´ç²—çš„ä¸»ç½‘æ ¼çº¿
    QPen mainGridPen(QColor(180, 180, 180, 200));
    mainGridPen.setWidth(2);
    painter.setPen(mainGridPen);
    
    // ç»˜åˆ¶ä¸»æ°´å¹³ç½‘æ ¼çº¿ï¼ˆæ¯éš”ä¸€ä¸ªç½‘æ ¼ï¼‰
    for (int y = m_logicalBounds.top(); y <= m_logicalBounds.bottom(); y += m_gridSize * 2) {
        QPoint start = logicalToScreen(QPoint(m_logicalBounds.left(), y));
        QPoint end = logicalToScreen(QPoint(m_logicalBounds.right(), y));
        painter.drawLine(start, end);
    }
    
    // ç»˜åˆ¶ä¸»å‚ç›´ç½‘æ ¼çº¿ï¼ˆæ¯éš”ä¸€ä¸ªç½‘æ ¼ï¼‰
    for (int x = m_logicalBounds.left(); x <= m_logicalBounds.right(); x += m_gridSize * 2) {
        QPoint start = logicalToScreen(QPoint(x, m_logicalBounds.top()));
        QPoint end = logicalToScreen(QPoint(x, m_logicalBounds.bottom()));
        painter.drawLine(start, end);
    }
    
    painter.restore();
}

/**
 * @brief ç»˜åˆ¶å±…æ°‘å°åŒº
 * @param painter ç»˜å›¾å¯¹è±¡
 */
void MainWindow::drawResidentialAreas(QPainter &painter)
{
    painter.save();
    
    painter.setPen(QPen(QColor(52, 152, 219), 4)); // æ›´ç²—çš„è“è‰²è¾¹æ¡†
    painter.setBrush(QBrush(QColor(173, 216, 230, 200))); // åŠé€æ˜æµ…è“è‰²å¡«å……
    
    for (const auto& area : m_locator.areas()) {
        QPoint screenPos = logicalToScreen(QPoint(area.x(), area.y()));
        
        // ç»˜åˆ¶å±…æ°‘å°åŒºï¼ˆåœ†å½¢ï¼‰- æ˜¾è‘—å¢å¤§å°ºå¯¸
        painter.drawEllipse(screenPos, 25, 25);
        
        // ç»˜åˆ¶å°åŒºæ ‡ç­¾
        painter.setPen(QPen(QColor(41, 128, 185), 2));
        QFont font = painter.font();
        font.setPointSize(10);
        font.setBold(true);
        painter.setFont(font);
        QString coordText = QString("(%1,%2)").arg(area.x()).arg(area.y());
        painter.drawText(screenPos + QPoint(30, -8), coordText);
        painter.setPen(QPen(QColor(52, 152, 219), 4));
    }
    
    painter.restore();
}

/**
 * @brief ç»˜åˆ¶é‚®å±€ä½ç½®
 * @param painter ç»˜å›¾å¯¹è±¡
 */
void MainWindow::drawPostOffice(QPainter &painter)
{
    if (!m_hasCalculated) return;
    
    painter.save();
    
    QPoint screenPos = logicalToScreen(m_optimalPostOffice);
    
    // ç»˜åˆ¶é‚®å±€å›¾æ ‡ï¼ˆçº¢è‰²åœ†å½¢ï¼‰- æ˜¾è‘—å¢å¤§å°ºå¯¸
    painter.setBrush(QBrush(QColor(231, 76, 60, 220))); // åŠé€æ˜çº¢è‰²å¡«å……
    painter.setPen(QPen(QColor(192, 57, 43), 3)); // æ›´ç²—çš„è¾¹æ¡†
    painter.drawEllipse(screenPos, 35, 35);
    
    // ç»˜åˆ¶é‚®å±€æ ‡ç­¾
    painter.setPen(QColor(192, 57, 43));
    QFont font = painter.font();
    font.setPointSize(12);
    font.setBold(true);
    painter.setFont(font);
    QString labelText = QString("ğŸ¤ é‚®å±€ (%1,%2)\nğŸ’° æ€»è´¹ç”¨: %3")
                        .arg(m_optimalPostOffice.x())
                        .arg(m_optimalPostOffice.y())
                        .arg(m_locator.calculateTotalCost(m_optimalPostOffice));
    painter.drawText(screenPos + QPoint(40, -15), labelText);
    
    painter.restore();
}

/**
 * @brief ç»˜åˆ¶è¿æ¥çº¿
 * @param painter ç»˜å›¾å¯¹è±¡
 */
void MainWindow::drawConnections(QPainter &painter)
{
    if (!m_hasCalculated) return;
    
    painter.save();
    
    QPen connectionPen(QColor(149, 165, 166, 180));
    connectionPen.setWidth(4); // å¢å¤§è¿æ¥çº¿å®½åº¦
    connectionPen.setStyle(Qt::DashLine);
    painter.setPen(connectionPen);
    
    QPoint postOfficeScreen = logicalToScreen(m_optimalPostOffice);
    
    for (const auto& area : m_locator.areas()) {
        QPoint areaScreen = logicalToScreen(QPoint(area.x(), area.y()));
        painter.drawLine(postOfficeScreen, areaScreen);
    }
    
    painter.restore();
}

/**
 * @brief é¼ æ ‡ç‚¹å‡»äº‹ä»¶å¤„ç†å‡½æ•°
 * @param event é¼ æ ‡äº‹ä»¶
 */
void MainWindow::mousePressEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        // æ£€æŸ¥ç‚¹å‡»æ˜¯å¦åœ¨ç»˜å›¾åŒºåŸŸå†…
        if (m_drawingArea.contains(event->pos())) {
            QPoint contentPos = event->pos() - m_drawingArea.topLeft();
            
            QPoint logicalPos = screenToLogical(contentPos);
            
            // è°ƒè¯•ä¿¡æ¯ï¼šæ˜¾ç¤ºåæ ‡è½¬æ¢ç»“æœ
            qDebug() << "é¼ æ ‡ç‚¹å‡»ä½ç½®:" << event->pos();
            qDebug() << "å†…å®¹åŒºåŸŸä½ç½®:" << contentPos;
            qDebug() << "é€»è¾‘åæ ‡ä½ç½®:" << logicalPos;
            
            // åˆ‡æ¢æ˜¾ç¤ºæ¨¡å¼ï¼šç‚¹å‡»ç»˜å›¾åŒºåŸŸæ—¶åœ¨XYåæ ‡å›¾å’Œç½‘æ ¼å›¾ä¹‹é—´åˆ‡æ¢
            m_showGridMode = !m_showGridMode;
            
            // æ·»åŠ å±…æ°‘å°åŒº
            QString areaName = QString("å°åŒº%1").arg(m_locator.areaCount() + 1);
            ResidentialArea newArea(logicalPos.x(), logicalPos.y(), areaName);
            m_locator.addArea(newArea);
            
            m_hasCalculated = false; // é‡ç½®è®¡ç®—çŠ¶æ€
            
            // ç›´æ¥æ›´æ–°ä¸»çª—å£æ˜¾ç¤º
            update();
            updateStatusBar();
        }
    }
    
    QMainWindow::mousePressEvent(event);
}

/**
 * @brief çª—å£å¤§å°æ”¹å˜äº‹ä»¶å¤„ç†å‡½æ•°
 * @param event å¤§å°æ”¹å˜äº‹ä»¶
 */
void MainWindow::resizeEvent(QResizeEvent *event)
{
    Q_UNUSED(event)
    
    // åŠ¨æ€è°ƒæ•´ç»˜å›¾åŒºåŸŸå¤§å°
    QSize windowSize = this->size();
    
    // è®¡ç®—å¯ç”¨ç»˜å›¾åŒºåŸŸå¤§å° - frameæ¡†ç°åœ¨åœ¨æœ€å·¦ä¾§
    int padding = 20;
    int availableWidth = windowSize.width() - ui->frame->width() - padding * 2; // å‡å»æ§åˆ¶é¢æ¿å®½åº¦å’Œè¾¹è·
    int availableHeight = windowSize.height() - 100; // å‡å»çŠ¶æ€æ å’Œè¾¹è·
    
    // è®¾ç½®ç»˜å›¾åŒºåŸŸçš„å°ºå¯¸ï¼Œç¡®ä¿ä¸å°äºæœ€å°å°ºå¯¸
    int drawingWidth = qMax(800, availableWidth);
    int drawingHeight = qMax(600, availableHeight);
    
    // æ›´æ–°ç»˜å›¾åŒºåŸŸçš„å®é™…å°ºå¯¸ - ä½äºframeæ¡†å³ä¾§
    m_drawingArea = QRect(ui->frame->width() + padding, padding, drawingWidth, drawingHeight);
    
    // é‡æ–°è®¡ç®—ç¼©æ”¾æ¯”ä¾‹
    m_scaleX = static_cast<double>(m_drawingArea.width()) / m_logicalBounds.width();
    m_scaleY = static_cast<double>(m_drawingArea.height()) / m_logicalBounds.height();
    
    // å¼ºåˆ¶é‡ç»˜
    update();
    updateStatusBar();
}

/**
 * @brief è®¡ç®—æœ€ä¼˜é‚®å±€ä½ç½®
 */
void MainWindow::calculateOptimalLocation()
{
    if (m_locator.areaCount() == 0) {
        QMessageBox::warning(this, "è­¦å‘Š", "è¯·å…ˆæ·»åŠ å±…æ°‘å°åŒºï¼");
        return;
    }
    
    m_optimalPostOffice = m_locator.findOptimalLocation();
    m_hasCalculated = true;
    
    // æ›´æ–°é€»è¾‘åæ ‡èŒƒå›´ä»¥åŒ…å«æ‰€æœ‰ç‚¹ï¼Œå¹¶æ·»åŠ è¾¹è·
    QRect range = m_locator.getCoordinateRange();
    int margin = 20; // è¾¹è·
    m_logicalBounds = QRect(range.left() - margin, range.top() - margin,
                           range.width() + 2 * margin, range.height() + 2 * margin);
    
    // é‡æ–°è®¡ç®—ç¼©æ”¾æ¯”ä¾‹
    m_scaleX = static_cast<double>(m_drawingArea.width()) / m_logicalBounds.width();
    m_scaleY = static_cast<double>(m_drawingArea.height()) / m_logicalBounds.height();
    
    // æ›´æ–°ä¸»çª—å£æ˜¾ç¤º
    update();
    updateStatusBar();
    
    // æ˜¾ç¤ºè®¡ç®—ç»“æœ
    int totalCost = m_locator.calculateTotalCost(m_optimalPostOffice);
    QMessageBox::information(this, "è®¡ç®—ç»“æœ", 
                            QString("æœ€ä¼˜é‚®å±€ä½ç½®: (%1, %2)\n"
                                   "æ€»è´¹ç”¨(è·ç¦»å’Œ): %3\n"
                                   "å±…æ°‘å°åŒºæ•°é‡: %4")
                            .arg(m_optimalPostOffice.x())
                            .arg(m_optimalPostOffice.y())
                            .arg(totalCost)
                            .arg(m_locator.areaCount()));
}

/**
 * @brief æ¸…ç©ºæ‰€æœ‰å±…æ°‘å°åŒº
 */
void MainWindow::clearAllAreas()
{
    if (m_locator.areaCount() == 0) {
        QMessageBox::information(this, "æç¤º", "å½“å‰æ²¡æœ‰å±…æ°‘å°åŒºå¯æ¸…ç©ºï¼");
        return;
    }
    
    int ret = QMessageBox::question(this, "ç¡®è®¤æ¸…ç©º", 
                                   "ç¡®å®šè¦æ¸…ç©ºæ‰€æœ‰å±…æ°‘å°åŒºå—ï¼Ÿ",
                                   QMessageBox::Yes | QMessageBox::No);
    
    if (ret == QMessageBox::Yes) {
        m_locator.clearAreas();
        m_hasCalculated = false;
        
        // é‡ç½®é€»è¾‘åæ ‡èŒƒå›´
        m_logicalBounds = QRect(-200, -200, 400, 400);
        
        // é‡æ–°è®¡ç®—ç¼©æ”¾æ¯”ä¾‹
        m_scaleX = static_cast<double>(m_drawingArea.width()) / m_logicalBounds.width();
        m_scaleY = static_cast<double>(m_drawingArea.height()) / m_logicalBounds.height();
        
        // æ›´æ–°ä¸»çª—å£æ˜¾ç¤º
        update();
        updateStatusBar();
    }
}

/**
 * @brief éšæœºç”Ÿæˆå±…æ°‘å°åŒº
 */
void MainWindow::generateRandomAreas()
{
    m_locator.clearAreas();
    
    QRandomGenerator *generator = QRandomGenerator::global();
    int areaCount = generator->bounded(5, 15); // ç”Ÿæˆ5-14ä¸ªå°åŒº
    
    for (int i = 0; i < areaCount; ++i) {
        int x = generator->bounded(-180, 180);
        int y = generator->bounded(-180, 180);
        QString name = QString("éšæœºå°åŒº%1").arg(i + 1);
        m_locator.addArea(ResidentialArea(x, y, name));
    }
    
    m_hasCalculated = false;
    
    // åˆ‡æ¢æ˜¾ç¤ºæ¨¡å¼ï¼šéšæœºç”Ÿæˆå°åŒºæ—¶åˆ‡æ¢åˆ°ç½‘æ ¼å›¾æ¨¡å¼
    m_showGridMode = true;
    
    // æ›´æ–°é€»è¾‘åæ ‡èŒƒå›´ä»¥åŒ…å«æ‰€æœ‰ç‚¹ï¼Œå¹¶æ·»åŠ è¾¹è·
    QRect range = m_locator.getCoordinateRange();
    int margin = 20; // è¾¹è·
    m_logicalBounds = QRect(range.left() - margin, range.top() - margin,
                           range.width() + 2 * margin, range.height() + 2 * margin);
    
    // é‡æ–°è®¡ç®—ç¼©æ”¾æ¯”ä¾‹
    m_scaleX = static_cast<double>(m_drawingArea.width()) / m_logicalBounds.width();
    m_scaleY = static_cast<double>(m_drawingArea.height()) / m_logicalBounds.height();
    
    // æ›´æ–°ä¸»çª—å£æ˜¾ç¤º
    update();
    updateStatusBar();
}



/**
 * @brief æ›´æ–°çŠ¶æ€æ ä¿¡æ¯
 */
void MainWindow::updateStatusBar()
{
    QString statusText = QString("å±…æ°‘å°åŒºæ•°é‡: %1 | ")
                        .arg(m_locator.areaCount());
    
    // æ·»åŠ æ˜¾ç¤ºæ¨¡å¼ä¿¡æ¯
    statusText += QString("æ˜¾ç¤ºæ¨¡å¼: %1 | ")
                 .arg(m_showGridMode ? "ç½‘æ ¼å›¾" : "XYåæ ‡å›¾");
    
    if (m_hasCalculated) {
        statusText += QString("æœ€ä¼˜é‚®å±€ä½ç½®: (%1, %2) | æ€»è´¹ç”¨: %3")
                     .arg(m_optimalPostOffice.x())
                     .arg(m_optimalPostOffice.y())
                     .arg(m_locator.calculateTotalCost(m_optimalPostOffice));
    } else {
        statusText += "è¯·ç‚¹å‡»'è®¡ç®—æœ€ä¼˜é‚®å±€ä½ç½®'æŒ‰é’®";
    }
    
    ui->statusbar->showMessage(statusText);
}

/**
 * @brief åæ ‡è½¬æ¢ï¼šé€»è¾‘åæ ‡åˆ°å±å¹•åæ ‡
 * @param logicalPoint é€»è¾‘åæ ‡ç‚¹
 * @return å±å¹•åæ ‡ç‚¹
 */
QPoint MainWindow::logicalToScreen(const QPoint &logicalPoint) const
{
    // å°†é€»è¾‘åæ ‡æ˜ å°„åˆ°å±å¹•åæ ‡
    double normalizedX = static_cast<double>(logicalPoint.x() - m_logicalBounds.left()) / m_logicalBounds.width();
    double normalizedY = static_cast<double>(logicalPoint.y() - m_logicalBounds.top()) / m_logicalBounds.height();
    
    int screenX = static_cast<int>(normalizedX * m_drawingArea.width());
    int screenY = static_cast<int>(normalizedY * m_drawingArea.height());
    
    // Yè½´ç¿»è½¬ï¼ˆå±å¹•åæ ‡Yè½´å‘ä¸‹ä¸ºæ­£ï¼‰
    screenY = m_drawingArea.height() - screenY;
    
    // è°ƒè¯•ä¿¡æ¯ï¼šæ˜¾ç¤ºåæ ‡è½¬æ¢ç»“æœ
    qDebug() << "ä¸»çª—å£åæ ‡è½¬æ¢ - é€»è¾‘åæ ‡:" << logicalPoint << "-> å±å¹•åæ ‡:" << QPoint(screenX, screenY);
    
    return QPoint(screenX, screenY);
}

/**
 * @brief åæ ‡è½¬æ¢ï¼šå±å¹•åæ ‡åˆ°é€»è¾‘åæ ‡
 * @param screenPoint å±å¹•åæ ‡ç‚¹
 * @return é€»è¾‘åæ ‡ç‚¹
 */
QPoint MainWindow::screenToLogical(const QPoint &screenPoint) const
{
    // Yè½´ç¿»è½¬ï¼ˆå±å¹•åæ ‡Yè½´å‘ä¸‹ä¸ºæ­£ï¼‰
    int flippedY = m_drawingArea.height() - screenPoint.y();
    
    // å°†å±å¹•åæ ‡æ˜ å°„åˆ°é€»è¾‘åæ ‡
    double normalizedX = static_cast<double>(screenPoint.x()) / m_drawingArea.width();
    double normalizedY = static_cast<double>(flippedY) / m_drawingArea.height();
    
    int logicalX = static_cast<int>(m_logicalBounds.left() + normalizedX * m_logicalBounds.width());
    int logicalY = static_cast<int>(m_logicalBounds.top() + normalizedY * m_logicalBounds.height());
    
    return QPoint(logicalX, logicalY);
}
